### PR TITLE
feature(reindex): add `include` arg

### PR DESCRIPTION
**Pull request by Marigold**
Reindexing of an entire catalog is slow (~10s) which is fine for bulk updates, but slow for fast-track. Add `include` argument to only reindex datasets matching given pattern and reuse all the other datasets from existing catalog.

### CODE REVIEW

File: owid/catalog/meta.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def pruned_json(cls: T) -> T:
     orig = cls.to_dict  # type: ignore
 
     # only keep non-null public variables
-    cls.to_dict = lambda self: {  # type: ignore
```
Comment (larsyencken) [inline owid/catalog/meta.py:22]: What's the purpose of this change?

File: owid/catalog/catalogs.py [hunk 2]
```diff
@@ -144,34 +145,43 @@ def _read_channels(self, channels: Iterable[CHANNEL]) -> pd.DataFrame:
...
+            if (dir / "index.json").exists() and re_search.search(str(dir)):
                 yield Dataset(dir)
                 continue
 
             for child in dir.iterdir():
                 if child.is_dir():
                     heapq.heappush(to_search, child)
 
-    def reindex(self) -> None:
+    def reindex(self, include: Optional[str] = None) -> None:
```
Comment (larsyencken) [inline owid/catalog/catalogs.py:167]: The logic seems correct, but honestly the `reindex()` method was already long and now it needs splitting. Could it look like this, at a high level?

```python
def reindex():
    index = _scan_for_datasets()
    _save_index(index)
    self.frame = index

def update_index(query):
    update = _scan_for_datasets(query)
    index = _merge_index(self.frame, update)
    _save_index(index)
    self.frame = index
```

File: owid/catalog/meta.py (hunk 1 above)
Comment (Marigold) [reply owid/catalog/meta.py:22]: Honestly, I don't remember. It fixed something when I was profiling (but eventually used a different approach). The problem is that `orig` is not assigned when you're creating the lambda (see [this example](https://stackoverflow.com/a/21054087/1275818)) and `lambda self, orig=orig: ...` guarantees it does.

Anyway, I can't replicate it now (could have been a mirage as well), so I'm putting the original back.

File: owid/catalog/catalogs.py (hunk 2 above)
Comment (Marigold) [reply owid/catalog/catalogs.py:167]: That's much better. Could you take a look now? I made some compromises but I'm not married to them.

File: owid/catalog/catalogs.py [hunk 3]
```diff
@@ -144,51 +145,81 @@ def _read_channels(self, channels: Iterable[CHANNEL]) -> pd.DataFrame:
...
+        """Walk the directory tree, generate a channel/namespace/version/dataset/table frame
+        and save it to feather."""
         self._save_metadata({"format_version": OWID_CATALOG_VERSION})
```
Comment (larsyencken) [inline owid/catalog/catalogs.py:170]: Move this line into `_save_index()`, and add a comment explaining it:

```python
# add a catalog version number that we can use to tell old clients to update
```

File: owid/catalog/catalogs.py [hunk 4]
```diff
@@ -144,51 +145,81 @@ def _read_channels(self, channels: Iterable[CHANNEL]) -> pd.DataFrame:
...
+        index = self._scan_for_datasets(include)
+
+        if include:
+            # we used regex to find datasets, so merge it with the original frame
+            index = self._merge_index(self.frame, index)
+
+        self._save_index(index)
+        self.frame = index
+        self.frame._base_uri = self.path.as_posix() + "/"
```
Comment (larsyencken) [inline owid/catalog/catalogs.py:179]: Move this line into the bottom of `_scan_for_datasets()`.

**Review by larsyencken — APPROVED** [review]
Much better! Just two more lines to hide in lower level methods, then ship it! :shipit:
